pulp_container/app/serializers.py

```diff
@@ -774,65 +772,80 @@ class OCIBuildImageSerializer(ValidateFieldsMixin, serializers.Serializer):
-                    _("Only one of 'containerfile' and 'containerfile_artifact' may be specified.")
-                )
-            data["containerfile_artifact"] = Artifact.init_and_validate(data.pop("containerfile"))
-        elif "containerfile_artifact" in data:
-            data["containerfile_artifact"].touch()
-        else:
+        if bool(data.get("containerfile", None)) == bool(data.get("containerfile_name", None)):
+            raise serializers.ValidationError(
+                _("Exactly one of 'containerfile' or 'containerfile_name' must be specified.")
+            )
+
+        if "containerfile_name" in data and "build_context" not in data:
             raise serializers.ValidationError(
-                _("'containerfile' or 'containerfile_artifact' must " "be specified.")
+                _("A 'build_context' must be specified when 'containerfile_name' is present.")
             )
-        artifacts = {}
-        if "artifacts" in data:
-            for url, relative_path in data["artifacts"].items():
-                if os.path.isabs(relative_path):
+
+        # the "has_repo_or_repo_ver_param_model_or_obj_perms" permission condition function
+        # expects a "repo" or "repository_version" arguments, so we need to pass "build_context"
+        # as "repository_version" to be able to validate the permissions
+        if data.get("build_context", None):
+            data["repository_version"] = data["build_context"]
+
+        return data
+
+    def deferred_validation(self, data):
+        """
+        Defer the validation of the `build_context` parameter to avoid rerunning unnecessary
+        database queries when checking permissions (DRF Access Policy).
+        """
+        if build_context := data.get("build_context", None):
+            data["content_artifact_pks"] = []
+            repository_version = RepositoryVersion.objects.get(pk=build_context.pk)
+            content_artifacts = ContentArtifact.objects.filter(
+                content__pulp_type="file.file", content__in=repository_version.content
+            ).order_by("-content__pulp_created")
+            for content_artifact in content_artifacts.select_related("artifact").iterator():
+                if not content_artifact.artifact:
                     raise serializers.ValidationError(
-                        _("Relative path cannot start with '/'. " "{0}").format(relative_path)
+                        _(
+                            "It is not possible to use File content synced with on-demand "
+                            "policy without pulling the data first."
+                        )
+                    )
+                containerfile_name = data.get("containerfile_name", None)
+                if containerfile_name and content_artifact.relative_path == containerfile_name:
+                    data["containerfile_artifact_pk"] = str(content_artifact.artifact.pk)
+
+                data["content_artifact_pks"].append(content_artifact.pk)
```

Comment:
When serializing data and dispatching a task from the viewset, there is a small window where someone could remove a file repository and we would end up with a list of non-existing keys. The task might fail when reading these keys. However, I am still unsure it is worth moving this code back to the task.